### DISCUSSION TITLE
SqlSyncChangeTrackingProvider and OnOutdated-Interceptor

### DISCUSSION THREAD

**Discussion by Christian-Fenzl** [Mimetis/Dotmim.Sync#1231] [Q&A] [answered]
Hi!

I'm using the SqlSyncChangeTrackingProvider on the server side to sync a database to local Sqlite databases and running into the following problem:

A record is inserted into a parent table with a corresponding record in a child table and successfully synced to the server database. Another client doesn't sync in time of the CHANGE_RENTATION days with the server and therefore did not get the this parent/child records when he syncs later on.

Then another client which is in sync, changes the child record and syncs. If the out of sync client now syncs within the CHANGE_RETENTION days for the change a sync error will be thrown, because it tries to inserts the child row without already having the parent record (FOREIGN KEY constraint failed).

So one problem is that after the CHANGE_RETENTION days there could be missing data if a client syncs too late and another problem is that the data is not consitent anymore and probably results in a sync error.

I was expecting that the OnOutdated-Interceptor will called if the last sync was too long ago for a client and the database will be completely recreated witht he following code:

```
 syncAgent.LocalOrchestrator.OnOutdated(args =>
 {
     args.Action = OutdatedAction.ReinitializeWithUpload;
 });
```

But this is not called. Do I miss something beside increasing the CHANGE_RETENTION days on the server? But this only would increase the time frame it works...

**Comment by Mimetis** [comment]
I think there is a problem with the way the `SqlSyncChangeTrackingProvider` is updating the last clean up timestamp.
I've made a correction in the last commit.

Until you have upgraded your version, you can try to manually update the `scope_info` table on the **server side**, by using this small script:

```

DECLARE @maxVersion int;
SELECT @maxVersion = MAX(CHANGE_TRACKING_MIN_VALID_VERSION(T.object_id)) 
FROM sys.tables T 
WHERE CHANGE_TRACKING_MIN_VALID_VERSION(T.object_id) is not null;

UPDATE scope_info SET sync_scope_last_clean_timestamp = @maxVersion;
```

Let me know if it's fixing your issue

**Reply by Christian-Fenzl** [reply]
Thanks for the fast help.

After applying your provided script on the server side the full update of the outdated local database is triggered as expected by calling my syncAgent.LocalOrchestrator.OnOutdated event handler and therefore lead to a full sync :)

I will update to version 1.2 (currently 0.9.8) and observe the value in scope_info.sync_scope_last_clean_timestamp in the future wether your adjustment prevents it from getting stale...

**Reply by Mimetis** [reply]
Nice ! 
Let me know the results of your further investigations, I'm interested in.

**Reply by Christian-Fenzl** [reply]
I tried to migrate from 0.9.8 to 1.2 but there is the exception "**sanity check on input stream failed, json token gap of more than 1048576 bytes**" thrown on the client side when the local database is not(!) yet available. Syncing an already existing database works so far.

Exception comes from here:
https://github.com/Mimetis/Dotmim.Sync/blob/c674550c033493602e9a0114ebcf3470a3de2a2f/Projects/Dotmim.Sync.Core/Serialization/JsonReader.cs#L171

What I tried:

- Limit the BatchSize to 50/500/1000 via the SyncOptions during the creation of the SyncAgent. No success.
- I downgraded back to version 1.0.0 and it worked again. 1.0.2 still works but from version 1.1.0 it breaks.

I'm using .Net 8 and the WebRemoteOrchestrator.

Any idea what changed from 1.1.0 on which could result in that?

**Reply by Mimetis** [reply]
are you using the same version on both side ? (`1.1.0` on client side and on server side) or are you using a different version on one of the side ? (like `1.1.0` on client and `0.9.8` on server side)

**Reply by Christian-Fenzl** [reply]
Yes, all projects are in the same solution and I updated all projects/dependencies to the same version.

**Reply by Mimetis** [reply]
Ok, Can you try (by using the code source directly) to remove the limitation in `JsonReader`:

``` cs
if (this.buffer.Length > MaxTokenGap)
  throw new JsonException($"sanity check on input stream failed, json token gap of more than {MaxTokenGap} bytes");
```

Just remove these two lines and let me know if it's working

**Reply by Christian-Fenzl** [reply]
I cloned version 1.2  and after removing the length check I copied the so modified "Dotmim.Sync.Core.dll" into my projects and it worked again :)

I tested with the modified dll:

- Syncing a not yet existing client database (which failed before)
- Syncing an existing and modified database
- Reprovisining on the server side.

**Reply by Mimetis** [reply]
Ok, So I will increase the buffer max length to `int.MaxValue` in the the next version
Thanks for the feedback !

**Reply by Christian-Fenzl** [reply]
Do you already know when you will release the next version approx.?

Thanks for your help and this great library! It's just great!

**Reply by Mimetis** [reply]
no Idea

**Reply by Christian-Fenzl** [reply]
> I think there is a problem with the way the `SqlSyncChangeTrackingProvider` is updating the last clean up timestamp. I've made a correction in the last commit.
> 
> Until you have upgraded your version, you can try to manually update the `scope_info` table on the **server side**, by using this small script:
> 
> ```
> 
> DECLARE @maxVersion int;
> SELECT @maxVersion = MAX(CHANGE_TRACKING_MIN_VALID_VERSION(T.object_id)) 
> FROM sys.tables T 
> WHERE CHANGE_TRACKING_MIN_VALID_VERSION(T.object_id) is not null;
> 
> UPDATE scope_info SET sync_scope_last_clean_timestamp = @maxVersion;
> ```
> 
> Let me know if it's fixing your issue

Maybe I'm wrong, but why do you need the **last** clean timestamp? Wouldn't it be better to check against the **current** one all he time instead of the one from the last sync? This would be more or less exactly what the script does when I applied it, which helped. With the added condition from your commit the value is updated less often, which leads to more missing synced records in my opinion because the needed tracking data can be removed already and this must be determined at the actual moment.

**Reply by Mimetis** [reply]
Indeed, but this is not how others providers works.
DMS is handling the last cleanup timestamp regardless the provider used.

But I think we can handling it in a better way, I guess.
Let me think a little bit about it, I will see if I can have a more accurate version of the last clean up, not when we are saving but also when we are needing it.

**Reply by Christian-Fenzl** [reply]
I understand. Would be much better to update the timestamp directly at the beginning before the checks, not at the end. Thanks for your time.

As an external workaround I can save my own last sync time on each client and if the next sync is later than the CHANGE_RETENTION days I can trigger a ReinitializeWithUpload instead of a Normal sync from the client side. But I would prefer that DMS can handle it truthfully by its own.

Please keep me informed. Until I can upgrade to the next version my workaround will be to manually apply the UPDATE script around every 10 days (CHANGE_RETENTION days is 14 days).

**Reply by Mimetis** [reply, marked as answer]
I've made a couple of changes in this [commit](https://github.com/Mimetis/Dotmim.Sync/commit/6c5bdaa797042f8ea8f6d7c77f80d31414442e2d) where I dynamically get the 
`MAX(CHANGE_TRACKING_MIN_VALID_VERSION)` of all tables, each time I try to get a `scope_info` record, that happens every time a sync starts.

**Reply by Christian-Fenzl** [reply]
This is great. Thank you so much!

Looking forward to the release of the next version...
